pkg/pki/k8scsrpki/k8scsr_user.go

```diff
@@ -0,0 +1,266 @@
+		if value, ok := secret.Data[v1alpha1.TLSJKSKeyStore]; !ok || len(value) == 0 {
+			jks, jksPasswd, err = certutil.GenerateJKS(certBundleX509, secret.Data[corev1.TLSPrivateKeyKey])
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	secret.Data[v1alpha1.TLSJKSKeyStore] = jks
+	secret.Data[v1alpha1.PasswordKey] = jksPasswd
+
+	typeMeta := secret.TypeMeta
+	err = c.client.Update(ctx, secret)
+	if err != nil {
+		return nil, err
+	}
+	secret.TypeMeta = typeMeta
+
+	return &pkicommon.UserCertificate{
+		CA:          secret.Data[corev1.TLSCertKey],
+		Certificate: secret.Data[corev1.TLSCertKey],
+		Key:         secret.Data[corev1.TLSPrivateKeyKey],
+		JKS:         secret.Data[v1alpha1.TLSJKSKeyStore],
+		Password:    secret.Data[v1alpha1.PasswordKey],
+	}, nil
+}
+
+// FinalizeUserCertificate removes/revokes a user certificate
+func (c *k8sCSR) FinalizeUserCertificate(_ context.Context, _ *v1alpha1.KafkaUser) error {
+	return nil
+}
+
+// getUserSigningRequest fetches the k8s signing request for a user
+func (c *k8sCSR) getUserSigningRequest(ctx context.Context, name, namespace string) (*certsigningreqv1.CertificateSigningRequest, error) {
+	signingRequest := &certsigningreqv1.CertificateSigningRequest{}
+	err := c.client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, signingRequest)
+	return signingRequest, err
+}
+
+func generateUserSecret(key []byte, secretName, namespace string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: key,
+		},
+	}
+}
+
+func generateCSRResource(csr []byte, name, namespace, signerName string,
+	annotation map[string]string) *certsigningreqv1.CertificateSigningRequest {
+	return &certsigningreqv1.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: name + "-",
+			Namespace:    namespace,
+			Annotations: util.MergeAnnotations(annotation,
+				map[string]string{pkicommon.KafkaUserAnnotationName: namespace + "/" + name}),
+		},
+		Spec: certsigningreqv1.CertificateSigningRequestSpec{
+			Request:    csr,
+			SignerName: signerName,
+			Usages:     []certsigningreqv1.KeyUsage{certsigningreqv1.UsageServerAuth, certsigningreqv1.UsageClientAuth},
+		},
+	}
+}
+
+func (c *k8sCSR) generateAndCreateCSR(ctx context.Context, clientkey []byte, user *v1alpha1.KafkaUser) (*certsigningreqv1.CertificateSigningRequest, error) {
+	c.logger.Info("Creating PKCS1PrivateKey from secret")
+	block, _ := pem.Decode(clientkey)
+	privKey, parseErr := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if parseErr != nil {
+		return nil, parseErr
+	}
+	c.logger.Info("Generating SigningRequest")
+	csr, err := certutil.GenerateSigningRequestInPemFormat(privKey, user.GetName(), []string{""})
+	if err != nil {
+		return nil, err
+	}
+	c.logger.Info("Generating k8s csr object")
+	signingReq := generateCSRResource(csr, user.GetName(), user.GetNamespace(),
```

Comment:
Using `user.GetName()` as name for CSR doesn't guarantee uniqueness cluster wide. (e.g. if there are two KafkaUser with the same name but different namespaces will map to the same CRS)

Reply:
That is right that why we are using generate name for the CSR, see https://github.com/banzaicloud/kafka-operator/blob/882ff6f80e5eee652b1e153c6ec55604b550cf65/pkg/pki/k8scsrpki/k8scsr_user.go#L219